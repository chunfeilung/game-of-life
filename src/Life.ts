/**
 * A grid holds information about a rectangular group of cells
 */
export type Grid = boolean[][];

/**
 * Simple, non-vectorised implementation of Conway’s Game of Life (also simply
 * called Life).
 *
 * Life is one of the most well-known examples of cellular automata. It has the
 * power of a universal Turing machine, which means that you could technically
 * use it for practical applications – that’s not what makes it so popular
 * however. Life also happens to be a fascinating simulation that shows how very
 * complex patterns can emerge from just a few simple rules.
 *
 * @example
 * let game = new Game(10, 10);
 * game.flip(1, 1);
 * game.flip(1, 2);
 * game.flip(2, 2);
 * let cells = game.evolve();
 */
export class Game {
  /**
   * Verify that an instance can be created using these parameter values.
   *
   * @param rows    Must be a positive integer.
   * @param columns Must be a positive integer.
   * @param seed    Must be able to contain values. Its size must not exceed
   *                the size of the grid.
   */
  private static checkConstructorParameters(rows: number,
                                            columns: number,
                                            seed: Grid,
  ): void {
    if (rows <= 0 || columns <= 0) {
      throw new Error("Row and column count must be positive integers");
    }

    if (seed !== null) {
      if (seed.length > rows || seed[0].length > columns) {
        throw new Error("Seed does not fit into grid");
      }
      if (seed.length === 0 || seed[0].length === 0) {
        throw new Error("Seed does not contain sufficient data");
      }
    }
  }

  /**
   * First, “blue” representation of the grid.
   */
  private readonly blueGrid: Grid;

  /**
   * Second, “green” representation of the grid.
   */
  private readonly greenGrid: Grid;

  /**
   * The pattern that is used to (re)initialise the grid.
   *
   * The seed pattern is optional, so this attribute can also be null.
   */
  private readonly seed: Grid;

  /**
   * The current representation of the grid.
   *
   * Alternates between the blue and green grid. Changes to individual cells
   * (via flip()) are applied to this grid.
   */
  private currentGrid: Grid;

  /**
   * The representation of the grid in the next iteration.
   *
   * Alternates between the blue and green grid. Changes made by evolving the
   * grid are applied to this grid.
   */
  private nextGrid: Grid;

  /**
   * Create a new instance of Game.
   *
   * @param rows    How many rows the grid should have; must be a positive
   *                integer. Very large values may cause performance issues.
   * @param columns How many columns the grid should have; must be a positive
   *                integer. Very large values may cause performance issues.
   * @param seed    Optional: the pattern used to initialise the grid.
   */
  public constructor(rows: number, columns: number, seed: Grid = null) {
    Game.checkConstructorParameters(rows, columns, seed);

    this.blueGrid = this.initialise(rows, columns);
    this.greenGrid = this.initialise(rows, columns);

    this.seed = seed;

    this.currentGrid = this.blueGrid;
    this.nextGrid = this.greenGrid;

    this.applySeedPattern();
  }

  /**
   * Toggle a cell between its active and inactive state.
   *
   * More specifically, this makes inactive (dead) cells active, and active
   * (alive) cells inactive. Surrounding cells won’t be affected until the next
   * iteration.
   *
   * @see {@link evolve} for advancing the simulation to the next iteration.
   *
   * @param row    Row of the cell that should be flipped. Zero-indexed.
   * @param column Column of the cell that should be flipped. Zero-indexed.
   */
  public flip = (row: number, column: number): Grid => {
    this.currentGrid[row][column] = !this.currentGrid[row][column];

    return this.currentGrid;
  }

  /**
   * Calculate the Game of Game instance’s next iteration.
   */
  public evolve = (): Grid => {
    // There’s a “current” grid and a “next” grid. This updates the “next” grid
    // first, and leaves the “current” grid untouched
    this.currentGrid.forEach((row: boolean[], rowIndex: number) => {
      row.forEach((_, columnIndex: number) => {
        const activeCells = this.countActiveCells(rowIndex, columnIndex);
        this.updateCell(rowIndex, columnIndex, activeCells);
      });
    });

    // until all cells in the “next” grid have been updated. This ensures that
    // clients will never see grids that are only partially updated
    this.switchGrids();

    return this.currentGrid;
  }

  /**
   * Revert the grid to its initial state.
   *
   * Will apply the seed pattern if one was passed during instance construction.
   */
  public reset = (): Grid => {
    this.currentGrid.map((row) => row.fill(false));

    this.applySeedPattern();

    return this.cells;
  }

  /**
   * Create the two-dimensional array that represents the grid.
   *
   * @param rows    How many rows the grid should have.
   * @param columns How many columns the grid should have.
   */
  private initialise = (rows: number, columns: number): Grid => {
    return Array(rows).fill(false).map(() => Array(columns).fill(false));
  }

  /**
   * Add the seed pattern to the grid, starting from the top-left corner.
   *
   * Will silently overwrite any values that were already present.
   */
  private applySeedPattern = (): void => {
    if (this.seed !== null) {
      this.seed.forEach((row: boolean[], rowIndex: number) => {
        row.forEach((cell: boolean, colIndex: number) => {
          this.currentGrid[rowIndex][colIndex] = cell;
        });
      });
    }
  }

  /**
   * The number of columns in the grid.
   */
  public get width(): number {
    return this.currentGrid[0].length;
  }

  /**
   * The number of rows in the grid.
   */
  public get height(): number {
    return this.currentGrid.length;
  }

  /**
   * Retrieve a representation of the complete grid as a two-dimensional array.
   */
  public get cells(): Grid {
    return this.currentGrid;
  }

  /**
   * Count the number of active cells in the direct vicinity of the current
   * cell.
   *
   * Includes the current cell itself!
   *
   * @param row    Row of the current cell.
   * @param column Column of the current cell.
   */
  private countActiveCells = (row: number, column: number): number => {
    let activeCells = 0;
    // Run a 3x3 kernel past every cell
    for (let i = row - 1; i <= row + 1; i++) {
      for (let j = column - 1; j <= column + 1; j++) {
        // Pretend that our grid is toroidal, i.e. the right and left, and
        // bottom and top edges are “stitched” together
        const wrappedRow = (i + this.height) % this.height;
        const wrappedCol = (j + this.width) % this.width;
        activeCells += Number(this.currentGrid[wrappedRow][wrappedCol]);
      }
    }
    return activeCells;
  }

  /**
   * Update a specific cell using a value that depends on the number of active
   * cells in its direct vicinity.
   *
   * @param row         Row of the cell that will be updated.
   * @param column      Column of the cell that will be updated.
   * @param activeCells Number of active cells in and around the current cell.
   */
  private updateCell = (row: number,
                        column: number,
                        activeCells: number): void => {
    // Conditions are very good, so the cell becomes alive (if it isn’t already)
    if (activeCells === 3) {
      this.nextGrid[row][column] = true;
      // Everything’s stable, nothing changes
    } else if (activeCells === 4) {
      this.nextGrid[row][column] = this.currentGrid[row][column];
      // Under- or overpopulation kills the cell
    } else {
      this.nextGrid[row][column] = false;
    }
  }

  /**
   * Makes the next iteration’s grid the current one.
   *
   * The current grid will be reused as the next iteration’s “next” grid.
   */
  private switchGrids = (): void => {
    const next = this.nextGrid;
    this.nextGrid = this.currentGrid;
    this.currentGrid = next;
  }
}
