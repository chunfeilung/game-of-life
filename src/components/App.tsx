import * as React from "react";
import * as Life from "../Life";
import {Button} from "./controls/Button";
import {Slider} from "./controls/Slider";
import {Grid} from "./Grid";

export interface IAppProps {
  game: Life.Game;
}

export interface IAppState {
  cells: Life.Grid;
  isPlaying: boolean;
  animation: NodeJS.Timer;
  speed: number;
}

/**
 * This is the main component.
 */
export class App extends React.Component<IAppProps, IAppState> {
  /**
   * Calculate the interval belonging to a certain speed setting.
   *
   * Intervals should be at least 1 second on the slowest setting, and decently
   * fast on the highest setting.
   *
   * @param speed How fast the simulation should be. Must be between 1 and 10.
   */
  private static determineInterval(speed: number): number {
    return 3000 / (3 * Math.pow(speed, 1.5));
  }

  /**
   * Initial state of the App.
   */
  public state = {
    animation: setInterval(null, 1000),
    cells: this.props.game.cells,
    isPlaying: false,
    speed: 5,
  };

  public render() {
    return (
      <div className="game-app">
        <div className="game-controls">
          <Button id="clearButton" label="Clear" onClick={this.clear}/>
          <Button id="nextButton" label="Next" onClick={this.next}/>
          <Button
            id="playButton"
            label={this.state.isPlaying ? "Pause" : "Play"}
            onClick={this.togglePlay}
          />
          <Slider
            id="slider"
            value={this.state.speed}
            onChange={this.adjustSpeed}
          />
        </div>
        <Grid cells={this.state.cells} onMouseDown={this.flip}/>
      </div>
    );
  }

  /**
   * Reverts the grid to its initial state.
   */
  private clear = (): void => {
    // Pressing “Clear” while the simulation is running should pause it
    if (this.state.isPlaying) {
      this.togglePlay();
    }

    this.props.game.reset();
    this.setState({cells: this.props.game.reset()});
  }

  /**
   * Advance the simulation to the next iteration.
   */
  private next = (): void => {
    // Pressing “Next” while the simulation is running should pause it
    if (this.state.isPlaying) {
      this.togglePlay();
    }
    this.evolve();
  }

  /**
   * Advance the simulation to the next iteration and update the grid.
   */
  private evolve = (): void => {
    this.props.game.evolve();
    this.setState({cells: this.props.game.cells});
  }

  /**
   * Automatically advance the simulation to next iterations.
   */
  private togglePlay = (): void => {
    const isPlaying = !this.state.isPlaying;
    this.setState({isPlaying});

    // Keep in mind that we’ve just updated isPlaying to the new state
    if (isPlaying) {
      this.setState({
        animation: setInterval(
          this.evolve,
          App.determineInterval(this.state.speed),
        ),
      });
    } else {
      clearInterval(this.state.animation);
    }
  }

  /**
   * Toggle a cell’s status between active and inactive.
   *
   * @param row    Row of the cell that should be toggled. Zero-indexed.
   * @param column Column of the cell that should be toggled. Zero-indexed.
   */
  private flip = (row: number, column: number) => () => {
    this.setState({cells: this.props.game.flip(row, column)});
  }

  /**
   * Updates the animation speed.
   *
   * @param speed How fast the animation should be. Must be between 1 and 10.
   */
  private adjustSpeed = (speed: number): void => {
    clearInterval(this.state.animation);
    this.setState({
      animation: setInterval(this.evolve, App.determineInterval(speed)),
      isPlaying: true,
    });
  }
}
