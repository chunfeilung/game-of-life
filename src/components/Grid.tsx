import * as React from "react";
import * as Life from "../Life";

export interface IGridProps {
  cells: Life.Grid;
  onMouseDown: (row: number, column: number) => () => void;
}

export const Grid = (props: IGridProps) => {
  /**
   * Render all cells in the grid.
   *
   * @param grid The grid that should be rendered.
   */
  const renderGrid = (grid: Life.Grid): JSX.Element[] => {
    return grid.map((row: boolean[], rowIndex: number) => {
      const rowCells = row.map((cell: boolean, columnIndex: number) => {
        return renderCell(rowIndex, columnIndex, cell);
      });
      return (<tr key={rowIndex}>{rowCells}</tr>);
    });
  };

  /**
   * Render a single cell.
   *
   * @param row    Row of the grid.
   * @param column Column of the grid.
   * @param value  Whether the cell is active or not.
   */
  const renderCell = (row: number,
                      column: number,
                      value: boolean): JSX.Element => {
    const key = row + "_" + column;
    return (
      <td
        id={"cell_" + key}
        key={key}
        className={"cell" + (value ? " -is-active" : "")}
        onMouseDown={props.onMouseDown(row, column)}>
        {value}
      </td>
    );
  };

  return (
    <table className="game-grid">
      <tbody>
      {renderGrid(props.cells)}
      </tbody>
    </table>
  );
};
