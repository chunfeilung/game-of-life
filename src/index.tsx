import * as React from "react";
import * as ReactDOM from "react-dom";
import {App} from "./components/App";
import {Game} from "./Life";
import "./styles/main.scss";

const game = new Game(60, 60);

// This initialises the game with an R-pentomino, a very simple pattern that
// evolves for quite a large number of iterations before it stabilises. By
// initialising the game using a series of manual flip()s, pressing “Clear” will
// still actually clear the grid.
//
// This behaviour might change in the future if I ever decide to add some
// predefined patterns with interesting properties.
game.flip(28, 29);
game.flip(28, 30);
game.flip(29, 28);
game.flip(29, 29);
game.flip(30, 29);

ReactDOM.render(
  <App game={game}/>,
  document.getElementById("game"),
);
