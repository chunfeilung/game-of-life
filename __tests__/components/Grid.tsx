import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import * as React from "react";
import * as Renderer from "react-test-renderer";
import {Grid} from "../../src/components/Grid";
import * as Life from "../../src/Life";

Enzyme.configure({adapter: new Adapter()});

const grid: Life.Grid = [
  [true, true, true],
  [false, false, false],
  [true, false, true],
  [false, true, false],
];

it("correctly renders a small grid", () => {
  /* Given */
  const component = Renderer.create(
    <Grid cells={grid} onMouseDown={jest.fn()}/>,
  );

  /* When */
  const snapshot = component.toJSON();

  /* Then */
  expect(snapshot).toMatchSnapshot();
});

it("provides the cellâ€™s coordinate to handler", () => {
  /* Given */
  // We define two variables. Values should be assigned later
  let row: number = null;
  let col: number = null;

  // This function will set the variables above
  const handleOnMouseDown = (y: number, x: number) => () => {
    row = y;
    col = x;
  };

  const component = Enzyme.shallow(
    <Grid cells={grid} onMouseDown={handleOnMouseDown}/>,
  );

  /* When */
  // Clicking the third cell on the second row
  const cell = component.find("#cell_1_2");
  cell.simulate("mouseDown");

  /* Then */
  // Variables should have been set to values that correspond with the
  // coordinates of the clicked cell
  expect(row).toBe(1);
  expect(col).toBe(2);
});
