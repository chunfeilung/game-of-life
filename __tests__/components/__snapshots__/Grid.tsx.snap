// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly renders a small grid 1`] = `
<table
  className="game-grid"
>
  <tbody>
    <tr>
      <td
        className="cell -is-active"
        id="cell_0_0"
      />
      <td
        className="cell -is-active"
        id="cell_0_1"
      />
      <td
        className="cell -is-active"
        id="cell_0_2"
      />
    </tr>
    <tr>
      <td
        className="cell"
        id="cell_1_0"
      />
      <td
        className="cell"
        id="cell_1_1"
      />
      <td
        className="cell"
        id="cell_1_2"
      />
    </tr>
    <tr>
      <td
        className="cell -is-active"
        id="cell_2_0"
      />
      <td
        className="cell"
        id="cell_2_1"
      />
      <td
        className="cell -is-active"
        id="cell_2_2"
      />
    </tr>
    <tr>
      <td
        className="cell"
        id="cell_3_0"
      />
      <td
        className="cell -is-active"
        id="cell_3_1"
      />
      <td
        className="cell"
        id="cell_3_2"
      />
    </tr>
  </tbody>
</table>
`;
